# ***Homework 02 -Data Mining Techniques***


Department of Informatics and Telecommunications University of Athens

Giorgos Pallikaris 1115201400335

Δ******* Γ*******       1115201900****

23/5/2o23 - 2/6/2o23

Grading 98/100

[Homework announcement ](https://drive.google.com/file/d/10qFXhdxX5dVjeY0IRcr_5DRPC3xzxXKv/view?usp=sharing)


In case you want to avoid running locally , a short image demonstration can be found [here ](https://drive.google.com/drive/folders/1AcTWyYwGN68RrGb0UVv97r8ns4OxfhSP?usp=sharing).



**Brief:** This homework is based on **Goodreads**.

Goodreads is a social platform with  a data base which contains numerous book data,mostly for commercialized books.

[https://www.goodreads.com/](https://)  .

During implementation we learned :   

**1.]  How to preprocess text that is going to be used in Natural Language Processing.**

**2.]  Differences and directions of count and prediction models ( non context based ).**

**3.]  We also refreshed a bit our M.L.  past experience as a tool to demonstrate our work on the above.**

**NOTES :**
 *Before first execution in your personal Jupyter Notebook
try finding keyword "CONTROL3".*

 *In the following cell delete the "#" right after comment1 and comment2 .*

*Have in mind that ,configurations can be done fast by finding the keyword "CONTROL" .*
  <br/><br/>
The .csv file can be found in zenodo repository : [https://zenodo.org/record/4265096#.Y-N2DnbP1jE](https://)  

The .csv file should be stored locally, at a location described by the following path : '/content/drive/MyDrive/DataMining/books_1.Best_Books_Ever.csv' .*

**For 1.ooo books:**


Time:  <1min


Space: <25GB

In [ ]:
from pandas.core.dtypes.common import conversion
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib import pyplot as plt
import csv

from mpl_toolkits import mplot3d
import sys

import numpy as np
from datetime import datetime
from datetime import timedelta
import sklearn
from sklearn import preprocessing
from sklearn import decomposition
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.metrics import davies_bouldin_score, silhouette_score
from scipy.spatial.distance import cdist
from sklearn.cluster import AgglomerativeClustering
from sklearn.metrics import silhouette_score

import matplotlib
%matplotlib inline
from google.colab import files
from ast import literal_eval

from google.colab import drive
drive.mount('/content/drive')

conversion = lambda x: literal_eval(str(x))
conv = {'genres': conversion,
        'characters': conversion,
        'awards': conversion,
        'ratingsByStars': conversion,
        'setting': conversion}

listofdates=["publishDate","firstPublishDate"]
path='/content/drive/MyDrive/DataMining/books_1.Best_Books_Ever.csv'




dataset1=pd.read_csv(path,converters=conv)
dataset1.head(3)


Mounted at /content/drive


bookId  \
0                     2767052-the-hunger-games   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                   2657.To_Kill_a_Mockingbird   

                                       title                 series  \
0                           The Hunger Games    The Hunger Games #1   
1  Harry Potter and the Order of the Phoenix        Harry Potter #5   
2                      To Kill a Mockingbird  To Kill a Mockingbird   

                                      author  rating  \
0                            Suzanne Collins    4.33   
1  J.K. Rowling, Mary GrandPré (Illustrator)    4.50   
2                                 Harper Lee    4.28   

                                         description language           isbn  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English  9780439023481   
1  There is a door at the end of a silent corrido...  English  9780439358071   
2  The unforgettable novel of a childhood in a sl...  English  9999999999999   

                                              genres  \
0  [Young Adult, Fiction, Dystopia, Fantasy, Scie...   
1  [Fantasy, Young Adult, Fiction, Magic, Childre...   
2  [Classics, Fiction, Historical Fiction, School...   

                                          characters  ... firstPublishDate  \
0  [Katniss Everdeen, Peeta Mellark, Cato (Hunger...  ...              NaN   
1  [Sirius Black, Draco Malfoy, Ron Weasley, Petu...  ...         06/21/03   
2  [Scout Finch, Atticus Finch, Jem Finch, Arthur...  ...         07/11/60   

                                              awards numRatings  \
0  [Locus Award Nominee for Best Young Adult Book...    6376780   
1  [Bram Stoker Award for Works for Young Readers...    2507623   
2  [Pulitzer Prize for Fiction (1961), Audie Awar...    4501075   

                              ratingsByStars likedPercent  \
0  [3444695, 1921313, 745221, 171994, 93557]         96.0   
1    [1593642, 637516, 222366, 39573, 14526]         98.0   
2  [2363896, 1333153, 573280, 149952, 80794]         95.0   

                                             setting  \
0  [District 12, Panem, Capitol, Panem, Panem (Un...   
1  [Hogwarts School of Witchcraft and Wizardry (U...   
2                 [Maycomb, Alabama (United States)]   

                                            coverImg  bbeScore bbeVotes  price  
0  https://i.gr-assets.com/images/S/compressed.ph...   2993816    30516   5.09  
1  https://i.gr-assets.com/images/S/compressed.ph...   2632233    26923   7.38  
2  https://i.gr-assets.com/images/S/compressed.ph...   2269402    23328    NaN  

[3 rows x 25 columns]

# **1.Preprocessing**

In [ ]:
print("Non numerical values per category\n\n",dataset1.isna().sum())

Non numerical values per category

 bookId                  0
title                   0
series              29008
author                  0
rating                  0
description          1338
language             3806
isbn                    0
genres                  0
characters              0
bookFormat           1473
edition             47523
pages                2347
publisher            3696
publishDate           880
firstPublishDate    21326
awards                  0
numRatings              0
ratingsByStars          0
likedPercent          622
setting                 0
coverImg              605
bbeScore                0
bbeVotes                0
price               14365
dtype: int64


1.1 **Cleaning**

Several values are missing or NaN.

In some fields these lines should be dropped to not affect negatively

the following ***Recommendation*** or ***Classification*** attempts.

In the rest fields missing elements either are valid or not to be used.

In [ ]:
dataset2=dataset1.copy()
dataset1=dataset2.dropna(subset=['bookId','ratingsByStars','genres','firstPublishDate','author','pages','publisher','description'])
dataset1=dataset1.reset_index(drop=True)
dataset1.isna().sum()


bookId                  0
title                   0
series              16536
author                  0
rating                  0
description             0
language             1149
isbn                    0
genres                  0
characters              0
bookFormat            293
edition             25833
pages                   0
publisher               0
publishDate             0
firstPublishDate        0
awards                  0
numRatings              0
ratingsByStars          0
likedPercent           75
setting                 0
coverImg              116
bbeScore                0
bbeVotes                0
price                4379
dtype: int64

1.2 **"Rating by stars" field expanding to specified ratings.**

In [ ]:

rati=['ratingStar']*5
i=[]
k=range(1,5)
fin=[]
for i in range(0,4):
    fin.append(str(rati[i])+str(k[i]))



pd.set_option('display.max_columns', None)

dfnew=pd.DataFrame(dataset1["ratingsByStars"].to_list(), columns=['ratingStar1','ratingStar2','ratingStar3','ratingStar4','ratingStar5'])


dataset1 = pd.concat([dataset1, dfnew], axis=1)
dataset1=dataset1.reset_index(drop=True)
dataset1=dataset1.dropna(subset=['ratingStar1'])
dataset1=dataset1.reset_index(drop=True)

dataset1.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28458 entries, 0 to 28457
Data columns (total 30 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   bookId            28458 non-null  object 
 1   title             28458 non-null  object 
 2   series            12090 non-null  object 
 3   author            28458 non-null  object 
 4   rating            28458 non-null  float64
 5   description       28458 non-null  object 
 6   language          27339 non-null  object 
 7   isbn              28458 non-null  object 
 8   genres            28458 non-null  object 
 9   characters        28458 non-null  object 
 10  bookFormat        28171 non-null  object 
 11  edition           2780 non-null   object 
 12  pages             28458 non-null  object 
 13  publisher         28458 non-null  object 
 14  publishDate       28458 non-null  object 
 15  firstPublishDate  28458 non-null  object 
 16  awards            28458 non-null  object

1.3 **Creating new -most weighted- genre field .**

In [ ]:
dataset1["genreSingle"] = dataset1["genres"].str[0]
dataset1=dataset1.dropna(subset=['genreSingle'])
#dataset1=dataset1.dropna(axis=0,how='any')
dataset1=dataset1.reset_index(drop=True)

"""
#lis=[]
#print(lis[0])
#print(dataset1.loc[[334]])
#lis=dataset1.genres.apply(lambda c: c==[])
#for i in range(len(lis)):
#  if lis[i]==True:
#    print(i)
"""
dataset1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27803 entries, 0 to 27802
Data columns (total 31 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   bookId            27803 non-null  object 
 1   title             27803 non-null  object 
 2   series            11983 non-null  object 
 3   author            27803 non-null  object 
 4   rating            27803 non-null  float64
 5   description       27803 non-null  object 
 6   language          26759 non-null  object 
 7   isbn              27803 non-null  object 
 8   genres            27803 non-null  object 
 9   characters        27803 non-null  object 
 10  bookFormat        27536 non-null  object 
 11  edition           2661 non-null   object 
 12  pages             27803 non-null  object 
 13  publisher         27803 non-null  object 
 14  publishDate       27803 non-null  object 
 15  firstPublishDate  27803 non-null  object 
 16  awards            27803 non-null  object

1.4 **New field creation for Year of first publish .**

In [ ]:
dataset1['newDate'] = pd.to_datetime(dataset1['firstPublishDate'],errors='coerce')
dataset1=dataset1.dropna(subset=['newDate'])


dataset1['publishingYear']=dataset1['newDate'].dt.year

def ye(x):
  if x['publishingYear']>2023:
    return x['publishingYear']-100
  else:
    return x['publishingYear']

dataset1['publishingYear'] = dataset1.apply(ye, axis=1)
dataset1['publishingYear'] = dataset1['publishingYear'].astype('int')

dataset1.head(10)

<ipython-input-6-a3d077763945>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset1['publishingYear']=dataset1['newDate'].dt.year


bookId  \
0  2.Harry_Potter_and_the_Order_of_the_Phoenix   
1                   2657.To_Kill_a_Mockingbird   
2                     1885.Pride_and_Prejudice   
3                               41865.Twilight   
4                         19063.The_Book_Thief   
5                           170448.Animal_Farm   
6               11127.The_Chronicles_of_Narnia   
7            30.J_R_R_Tolkien_4_Book_Boxed_Set   
8                     18405.Gone_with_the_Wind   
9  386162.The_Hitchhiker_s_Guide_to_the_Galaxy   

                                               title  \
0          Harry Potter and the Order of the Phoenix   
1                              To Kill a Mockingbird   
2                                Pride and Prejudice   
3                                           Twilight   
4                                     The Book Thief   
5                                        Animal Farm   
6                           The Chronicles of Narnia   
7  J.R.R. Tolkien 4-Book Boxed Set: The Hobbit an...   
8                                 Gone with the Wind   
9               The Hitchhiker's Guide to the Galaxy   

                                              series  \
0                                    Harry Potter #5   
1                              To Kill a Mockingbird   
2                                                NaN   
3                               The Twilight Saga #1   
4                                                NaN   
5                                                NaN   
6  The Chronicles of Narnia (Publication Order) #1–7   
7                         The Lord of the Rings #0-3   
8                                                NaN   
9            The Hitchhiker's Guide to the Galaxy #1   

                                              author  rating  \
0          J.K. Rowling, Mary GrandPré (Illustrator)    4.50   
1                                         Harper Lee    4.28   
2          Jane Austen, Anna Quindlen (Introduction)    4.26   
3                                    Stephenie Meyer    3.60   
4                    Markus Zusak (Goodreads Author)    4.37   
5  George Orwell, Russell Baker (Preface), C.M. W...    3.95   
6           C.S. Lewis, Pauline Baynes (Illustrator)    4.26   
7                                     J.R.R. Tolkien    4.60   
8                                  Margaret Mitchell    4.30   
9                                      Douglas Adams    4.22   

                                         description language           isbn  \
0  There is a door at the end of a silent corrido...  English  9780439358071   
1  The unforgettable novel of a childhood in a sl...  English  9999999999999   
2  Alternate cover edition of ISBN 9780679783268S...  English  9999999999999   
3  About three things I was absolutely positive.\...  English  9780316015844   
4  Librarian's note: An alternate cover edition c...  English  9780375831003   
5  Librarian's note: There is an Alternate Cover ...  English  9780451526342   
6  Journeys to the end of the world, fantastic cr...  English  9999999999999   
7  This four-volume, boxed set contains J.R.R. To...  English  9780345538376   
8  Scarlett O'Hara, the beautiful, spoiled daught...  English  9780446675536   
9  Seconds before the Earth is demolished to make...  English  9999999999999   

                                              genres  \
0  [Fantasy, Young Adult, Fiction, Magic, Childre...   
1  [Classics, Fiction, Historical Fiction, School...   
2  [Classics, Fiction, Romance, Historical Fictio...   
3  [Young Adult, Fantasy, Romance, Vampires, Fict...   
4  [Historical Fiction, Fiction, Young Adult, His...   
5  [Classics, Fiction, Dystopia, Fantasy, Literat...   
6  [Fantasy, Classics, Fiction, Young Adult, Chil...   
7  [Fantasy, Fiction, Classics, Adventure, Scienc...   
8  [Classics, Historical Fiction, Fiction, Romanc...   
9  [Science Fiction, Fiction, Humor, Fantasy, Cla...   

                                          characters

In [ ]:
dataset1.isna().sum()

bookId                  0
title                   0
series              15808
author                  0
rating                  0
description             0
language             1043
isbn                    0
genres                  0
characters              0
bookFormat            267
edition             25130
pages                   0
publisher               0
publishDate             0
firstPublishDate        0
awards                  0
numRatings              0
ratingsByStars          0
likedPercent            0
setting                 0
coverImg               81
bbeScore                0
bbeVotes                0
price                4033
ratingStar1             0
ratingStar2             0
ratingStar3             0
ratingStar4             0
ratingStar5             0
genreSingle             0
newDate                 0
publishingYear          0
dtype: int64

# **2. Simple Questions to dive into data's nature**

**5 out of 13 will be**: 2 , 3  , 5 , 9 , 13



**q2.10 books with the most pages**

In [ ]:
data1=pd.DataFrame()

dataset1['pages']=dataset1.pages.str.extract('(\d+)')
dataset1['pages'] = dataset1['pages'].astype(int)
data1=dataset1.nlargest(10, ["pages"])


print(data1.title.to_string(index=False))

                     The Story of Civilization
                                     Animorphs
                                          Worm
                           سه تفنگدار: ١٠ جلدی
The Complete Aubrey/Maturin Novels (5 Volumes)
                              البداية والنهاية
                        A Song of Ice and Fire
        Cuentos completos 1880 -1903 [Estuche]
                          The Second World War
                             سير أعلام النبلاء


**q3. 10 most 5-starred books**

In [ ]:
dataset1['ratingStar5'] = dataset1['ratingStar5'].astype(str).astype(int)
dataset1.nlargest(10, ["ratingStar5"])

bookId  \
3                             41865.Twilight   
35               5107.The_Catcher_in_the_Rye   
20                    7624.Lord_of_the_Flies   
27   3.Harry_Potter_and_the_Sorcerer_s_Stone   
144                      19501.Eat_Pray_Love   
24                     4671.The_Great_Gatsby   
22                    18144590-the-alchemist   
417                        960.Angels_Demons   
5                         170448.Animal_Farm   
1                 2657.To_Kill_a_Mockingbird   

                                     title                 series  \
3                                 Twilight   The Twilight Saga #1   
35                  The Catcher in the Rye                    NaN   
20                       Lord of the Flies                    NaN   
27   Harry Potter and the Sorcerer's Stone        Harry Potter #1   
144                        Eat, Pray, Love                    NaN   
24                        The Great Gatsby                    NaN   
22                           The Alchemist                    NaN   
417                        Angels & Demons      Robert Langdon #1   
5                              Animal Farm                    NaN   
1                    To Kill a Mockingbird  To Kill a Mockingbird   

                                                author  rating  \
3                                      Stephenie Meyer    3.60   
35                                       J.D. Salinger    3.81   
20                                     William Golding    3.69   
27           J.K. Rowling, Mary GrandPré (Illustrator)    4.47   
144               Elizabeth Gilbert (Goodreads Author)    3.57   
24       F. Scott Fitzgerald, Francis Scott Fitzgerald    3.92   
22   Paulo Coelho (Goodreads Author), Alan R. Clark...    3.88   
417                       Dan Brown (Goodreads Author)    3.90   
5    George Orwell, Russell Baker (Preface), C.M. W...    3.95   
1                                           Harper Lee    4.28   

                                           description language  \
3    About three things I was absolutely positive.\...  English   
35   The hero-narrator of The Catcher in the Rye is...  English   
20   At the dawn of the next world war, a plane cra...  English   
27   Harry Potter's life is miserable. His parents ...  English   
144  A celebrated writer's irresistible, candid, an...  English   
24   Alternate Cover Edition ISBN: 0743273567 (ISBN...  English   
22   Paulo Coelho's enchanting novel has inspired a...  English   
417  World-renowned Harvard symbologist Robert Lang...  English   
5    Librarian's note: There is an Alternate Cover ...  English   
1    The unforgettable novel of a childhood in a sl...  English   

              isbn                                             genres  \
3    9780316015844  [Young Adult, Fantasy, Romance, Vampires, Fict...   
35   9780316769174  [Classics, Fiction, Young Adult, Literature, S...   
20   9780140283334  [Classics, Fiction, Young Adult, School, Liter...   
27   9999999999999  [Fantasy, Fiction, Young Adult, Magic, Childre...   
144  9780143038412  [Nonfiction, Memoir, Travel, Biography, Chick ...   
24   9999999999999  [Classics, Fiction, School, Literature, Histor...   
22   9780062315007  [Fiction, Classics, Fantasy, Philosophy, Novel...   
417  9781416524793  [Fiction, Mystery, Thriller, Suspense, Mystery...   
5    9780451526342  [Classics, Fiction, Dystopia, Fantasy, Literat...   
1    9999999999999  [Classics, Fiction, Historical Fiction, School...   

                                            characters             bookFormat  \
3    [Edward Cullen, Jacob Black, Laurent, Renee, B...              Paperback   
35   [Holden Caulfield, Robert Ackley, Stradlater, ...              Paperback   
20         [Ralph, Piggy, Roger, Jack Merridew, Simon]              Paperback   
27   [Draco Malfoy, Ron Weasley, Petunia Dursley, V...              Hardcover   
144   [Liz Gilbert, Felipe, Richard from Texas, Wayan]              Paperba

**q5. 10 writers with the most published books**

In [ ]:
#pd.set_option("display.max_rows", 10)
dataset1['author'].value_counts().head(10)

Agatha Christie                       68
Terry Pratchett                       43
Nora Roberts (Goodreads Author)       41
Stephen King (Goodreads Author)       40
Enid Blyton                           40
Dean Koontz (Goodreads Author)        38
Karen Kingsbury (Goodreads Author)    36
P.G. Wodehouse                        34
Orson Scott Card                      34
Bonnie Bryant                         33
Name: author, dtype: int64

**q9.10 publishers with the most published books**

In [ ]:
dataset1['publisher'].value_counts().head(10)

Vintage             549
Penguin Books       400
HarperCollins       305
Ballantine Books    298
Bantam              284
Penguin Classics    271
Pocket Books        232
Del Rey             218
Harper Perennial    217
Tor Books           191
Name: publisher, dtype: int64

**q13.Books published per year**

In [ ]:
pd.set_option("display.max_rows",None)

dataset1['publishingYear'].value_counts(sort=True)

2010    1090
2011    1079
2012    1059
2009    1055
2013     997
2004     911
2005     904
2006     895
2003     852
2008     846
2007     813
2014     750
2002     743
2000     684
2001     676
1999     581
1998     573
1997     528
1996     523
1995     484
1994     445
2015     425
1992     412
1993     401
1991     362
1990     359
1989     326
1988     325
1987     323
1985     288
1986     286
1983     275
1984     271
1981     229
2016     221
1980     214
1982     212
1977     197
1978     197
1975     194
1976     181
1979     169
1974     163
1973     155
1972     152
1970     150
2017     148
1969     144
1971     144
1959     138
1962     135
1968     130
1961     129
1963     129
1966     126
2018     124
1964     122
1955     121
1967     117
1965     117
1957     115
1950     113
1958     109
1953     107
1960     104
1956     103
1951     103
2019      98
1954      97
1949      91
1939      89
1947      88
1946      79
1948      78
1952      77
1944      77
1937      76

# **3. Recommendation System**

Our 3 main targets are :
1. Predictions of a user's review on a read book.
2. Display of a sorted list of N books that a user would like to knew more.
3. Assist user to discover books that would not have the chance to discover otherways.




For the above puproses , TF-IDF  unigram and bigram will be used
based on  **"description"** field.

Cosine similarity will be the metric.

Let's start by keeping only entries with description written in English.

In [ ]:
dataset1=dataset1[dataset1['language']=='English']
dataset1.shape

(23758, 33)

3.1 Prediction

In [ ]:
dataframe1 = dataset1[['bookId', 'description']].copy()
dataframe1.shape

(23758, 2)

Our task need O(n^2) time thus at least c*(23758 ^2) which is more than 1 hour in this executional enviroment.

For time economy reasons we will work with a much smaller dataset ,with 1000 rows.
Later we ll have the choice to run our recommendation algorithm for the whole dataset.

In [ ]:
####### CONTROL BOX ##################
books_quantity=1000

######################################



def strip_everything(ww):
  w1 = ww.strip("'-:()[]{}!?$%^&*_~\;/.`+=#@0123456789")
  return w1



dataframe2=dataframe1.iloc[:books_quantity].copy()

dataframe2.info


import nltk
import re
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords
nltk.download('punkt')

nltk.download('stopwords')
nltk.download('wordnet')
stopwords = nltk.corpus.stopwords.words('english')
#stopwords.extend( ['and', 'didnt',])
lemmatizer = WordNetLemmatizer()

# for each row
for jj in  range (len( dataframe2  ) ):


  cour= dataframe2.iloc[jj,1]
  if books_quantity<51:
    print ( cour,"\n")





  cour = cour.replace("−", " 9 ")
  cour = re.sub(r'-' ,' 9 ',cour);

  ######  Important ** The sequence of our cleaning ,converting actions affects
  #####   the final result.Thus we  have to be carefoul on possible conflicts
  #### i.e. bla bla bla - blu1 blu blu blu2 - ble ble ble
  ### may become bla bla blablu1 ..... blu2ble ble ble
  # instead of bla bla bla blu1 blu blu blu2 ble ble ble


  #####  A regular expression way to delete everything that is not a w ( \w)
  #####  or whitespace (\s).Negation is with  symbol :  ^
  #####  \d for  digiits [0-9]
  #####  | == OR
  #####  \'\w+ for letter after '

  cour = re.sub(r'\'\w+', '', cour)
  cour  =re.sub(r'[^\w\s]|\d', '', cour)

  #a lambda calculus way for the same purpose
  #cour=' '.join(list(map(lambda x: x[:x.find("'")] if "'" in x else x, cour.split(' '))))

  #convert every upper case character to it's lower case
  cour = cour.lower()

  #deleting parenthesis and non alphanumerical characters
  cour = strip_everything(cour)

  #convert text to list/vector
  #cour =word_tokenize(cour)
  if books_quantity<51:
    print ( cour,"\n")

  """
  filtered_sentence = [w for w in word_tokens if not w.lower() in stopwords]

  for w in word_tokens:
    if w not in stopwords:
        filtered_sentence.append(w)

  """


  lemmatizer= WordNetLemmatizer()
  """for kk in range  ( len ( cour)):
   #print ( filtered_sentence[kk])
   #print ( lemmatizer.lemmatize(filtered_sentence[kk]))
   lemmed_result= lemmatizer.lemmatize(cour[kk])
  """

  def textTolemma(wordtext):
    wordss=wordtext.split()
    resulttext=[]
    for wor in wordss:
      resulttext.append(lemmatizer.lemmatize(wor))

    return ' '.join(resulttext)




  dataframe2.at[jj,'description']=cour


dataframe2.head()

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


bookId  \
0  2.Harry_Potter_and_the_Order_of_the_Phoenix   
1                   2657.To_Kill_a_Mockingbird   
2                     1885.Pride_and_Prejudice   
3                               41865.Twilight   
4                         19063.The_Book_Thief   

                                         description  
0  there is a door at the end of a silent corrido...  
1  the unforgettable novel of a childhood in a sl...  
2  alternate cover edition of isbn since its imme...  
3  about three things i was absolutely positive\n...  
4  librarian note an alternate cover edition can ...

For unigram :

In [ ]:
vectorizer1=TfidfVectorizer( lowercase=True,max_df=0.8, min_df=0.05,ngram_range=(1,1),stop_words= 'english')



vect_res1 = vectorizer1.fit_transform(dataframe2['description'])
feature_names1= vectorizer1.get_feature_names_out()
dense1=vect_res1.todense()
denselist1=dense1.tolist()
df_res1=pd.DataFrame(denselist1,columns=feature_names1)



list_ofVector_results1=[]
for kk in range ( len ( df_res1)):
  temp=df_res1.iloc[kk].to_numpy()
  #now we have a list of vectorized texts which we will insert in dataframe
  list_ofVector_results1.append(temp)

dataframe2['Unigram_vectors']=list_ofVector_results1

print( dataframe2)


                                                 bookId  \
0           2.Harry_Potter_and_the_Order_of_the_Phoenix   
1                            2657.To_Kill_a_Mockingbird   
2                              1885.Pride_and_Prejudice   
3                                        41865.Twilight   
4                                  19063.The_Book_Thief   
5                                    170448.Animal_Farm   
6                        11127.The_Chronicles_of_Narnia   
7                     30.J_R_R_Tolkien_4_Book_Boxed_Set   
8                              18405.Gone_with_the_Wind   
9           386162.The_Hitchhiker_s_Guide_to_the_Galaxy   
10                               370493.The_Giving_Tree   
11                               6185.Wuthering_Heights   
12                                968.The_Da_Vinci_Code   
13                              929.Memoirs_of_a_Geisha   
14                      5297.The_Picture_of_Dorian_Gray   
15    24213.Alice_s_Adventures_in_Wonderland_Through... 

For bigram :

In [ ]:
vectorizer2=TfidfVectorizer( lowercase=True,max_df=0.9, min_df=0.05,ngram_range=(2,2),stop_words= 'english')


vect_res2 = vectorizer2.fit_transform(dataframe2['description'])
feature_names2= vectorizer2.get_feature_names_out()
dense2=vect_res2.todense()
denselist2=dense2.tolist()
df_res2=pd.DataFrame(denselist2,columns=feature_names2)


list_ofVector_results2=[]
for kk in range ( len ( df_res2)):
  temp=df_res2.iloc[kk].to_numpy()
  #now we have a list of vectorized texts which we will insert in dataframe
  list_ofVector_results2.append(temp)

dataframe2['Bigram_vectors']=list_ofVector_results2

print( dataframe2)

                                                 bookId  \
0           2.Harry_Potter_and_the_Order_of_the_Phoenix   
1                            2657.To_Kill_a_Mockingbird   
2                              1885.Pride_and_Prejudice   
3                                        41865.Twilight   
4                                  19063.The_Book_Thief   
5                                    170448.Animal_Farm   
6                        11127.The_Chronicles_of_Narnia   
7                     30.J_R_R_Tolkien_4_Book_Boxed_Set   
8                              18405.Gone_with_the_Wind   
9           386162.The_Hitchhiker_s_Guide_to_the_Galaxy   
10                               370493.The_Giving_Tree   
11                               6185.Wuthering_Heights   
12                                968.The_Da_Vinci_Code   
13                              929.Memoirs_of_a_Geisha   
14                      5297.The_Picture_of_Dorian_Gray   
15    24213.Alice_s_Adventures_in_Wonderland_Through... 

Cosine similarity metric will be used to estimate similarities between description's.



In [ ]:
barrier100=100



from sklearn.metrics.pairwise import cosine_similarity
from numpy import dot
from numpy.linalg import norm
import math
from scipy import spatial

countera=0

final_dict_uni={}
for i in range ( len ( dataframe2)):
  list_100_similar=[]
  for k in range ( len ( dataframe2)):
    countera+=1
    if i==k:
      continue
    else:
      #print ( " Book 1 id = " ,dataframe2['bookId'].iloc[i])

      a = dataframe2['Unigram_vectors'].iloc[i]
      b=dataframe2['Unigram_vectors'].iloc[k]
      #c=dataframe2['description'].iloc[k]

      cos_sim=dot(a, b)/(norm(a)*norm(b))
      if math.isnan(cos_sim):
        cos_sim=0.0
      #print ( cos_sim,"\n\n")

      sort_book_entry=[dataframe2['bookId'].iloc[k], cos_sim ]
      #print ( " Is geometrically close with ", sort_book_entry,"\n\n")


      ###### Inserting in the list which will become the dictionary #####

      #if best 100 list is empty insert the first elemnt you examine
      if len(list_100_similar)==0:
        list_100_similar.append(sort_book_entry)

      else:#if it is not empty instert the element you examine in the correct position

        #SORTINSERT
        for s in range( len(list_100_similar) ):

          # if it is smaller than current index
          if list_100_similar[s][1] >=sort_book_entry[1]:

            indy=s

            # if it is not full yet add everyone
            if len(list_100_similar) <barrier100:
              list_100_similar=list_100_similar[:indy] + [sort_book_entry] + list_100_similar[indy:]

              break


            #if it is Full you have to add in correct position
            else:
              #if it is smaller than the smallest
              #if list_100_similar[0][1] >sort_book_entry[1] :
              if s==0:

                break
              #if it smaller than the current and the current is not the smallest
              # add it here
              else:

                list_100_similar=list_100_similar[1:indy] + [sort_book_entry] + list_100_similar[indy:]
                break


          # if it is bigger than current index
          else :


            #if it is the end of the list
            if s == len(list_100_similar)-1:

              #add it and remove the first one if it is foul
              if s==barrier100-1:

                list_100_similar=list_100_similar[1:s]+ [sort_book_entry]
                break
              # add it at the end if it is not foul
              else :

                list_100_similar=list_100_similar + [sort_book_entry]
                break

            else:

              continue


  list_100_similar.reverse()
  final_dict_uni.update ({ dataframe2['bookId'].iloc[i] : list_100_similar })
  #print (len (list_100_similar), list_100_similar, '\n\n\n')


"""
for i in final_dict_uni:
  print ( " book is : ",i, '\n and the most similar books in increasing order are ', final_dict_uni[i] , '\n\n ')
"""

<ipython-input-18-3ef2e98c1af7>:27: RuntimeWarning: invalid value encountered in double_scalars
  cos_sim=dot(a, b)/(norm(a)*norm(b))


'\nfor i in final_dict_uni:\n  print ( " book is : ",i, \'\n and the most similar books in increasing order are \', final_dict_uni[i] , \'\n\n \')\n'

In [ ]:
"""
This function is used to recomend up to 100 books from the known base
that are consider similar to a given books id
"""
def book_recommender ( book_id="19063.The_Book_Thief" ,quantity=5,book_dict=final_dict_uni,dfvar=dataframe1,max_books=books_quantity):
  print ("Recommending ",quantity," books similar to:" ,book_id,"\n------------------------------------------------------\n")
  similar_book_dec_list=book_dict[book_id]

  postplus=0
  for i in range ( quantity):
    """
    if pd.isna( similar_book_dec_list[i][1] ) :
      postplus+=1

  quantity+=postplus
  """

  for i in range (quantity):

    if pd.isna( similar_book_dec_list[i][1] ) :
      i+=1
      continue

    sugid=similar_book_dec_list[i][0]
    org_row= dfvar.loc[dfvar['bookId'] == sugid]
    str_desc=org_row['description']

    t=len(str(max_books))
    t=t+3


    str_desc=str_desc.to_string()
    red_str=str_desc[t : ]

    if similar_book_dec_list[i][1]==0:
      print ( " Sorry \nOnly " ,i+1," books overcome the lowest threshold  of similarity with ",book_id,'!')
      break
    print("Recommended : ", sugid)
    print("Description : ",red_str)
    print (" score : ",similar_book_dec_list[i][1], '\n\n')



book_recommender("2.Harry_Potter_and_the_Order_of_the_Phoenix",90 )

Recommending  90  books similar to: 2.Harry_Potter_and_the_Order_of_the_Phoenix 
------------------------------------------------------

Recommended :  31122.I_Capture_the_Castle
Description :  Through six turbulent months of 1934, 17-year-...
 score :  0.48463490837338163 


Recommended :  18521.A_Room_of_One_s_Own
Description :  A Room of One's Own is an extended essay by Vi...
 score :  0.4233528886559716 


Recommended :  11507.My_Name_Is_Asher_Lev
Description :  Asher Lev is a Ladover Hasid who keeps kosher,...
 score :  0.38974214082538 


Recommended :  119073.The_Name_of_the_Rose
Description :  The year is 1327. Benedictines in a wealthy It...
 score :  0.37289630021529535 


Recommended :  15997.Paradise_Lost
Description :  John Milton's Paradise Lost is one of the grea...
 score :  0.3642583814592007 


Recommended :  3388.Corelli_s_Mandolin
Description :  Captain Corelli’s Mandolin is set in the early...
 score :  0.35543348511330536 


Recommended :  29797.The_Pilgrim_s_Pro

In [ ]:
"""
This function is used to recomend up to 100 books from the known base
that are consider similar to a given books id
"""
def book_recommender ( book_id="19063.The_Book_Thief" ,quantity=5,book_dict=final_dict_uni,dfvar=dataframe1,max_books=books_quantity):
  print ("Recommending ",quantity," books similar to:" ,book_id,"\n------------------------------------------------------\n")
  similar_book_dec_list=book_dict[book_id]

  postplus=0
  for i in range ( quantity):
    """
    if pd.isna( similar_book_dec_list[i][1] ) :
      postplus+=1

  quantity+=postplus
  """

  for i in range (quantity):

    if pd.isna( similar_book_dec_list[i][1] ) :
      i+=1
      continue

    sugid=similar_book_dec_list[i][0]
    org_row= dfvar.loc[dfvar['bookId'] == sugid]
    str_desc=org_row['description']

    t=len(str(max_books))
    t=t+3


    str_desc=str_desc.to_string()
    red_str=str_desc[t : ]

    if similar_book_dec_list[i][1]==0:
      print ( " Sorry \nOnly " ,i+1," books overcome the lowest threshold  of similarity with ",book_id,'!')
      break
    print("Recommended : ", sugid)
    print("Description : ",red_str)
    print (" score : ",similar_book_dec_list[i][1], '\n\n')



book_recommender("2.Harry_Potter_and_the_Order_of_the_Phoenix",90 )

Recommending  90  books similar to: 2.Harry_Potter_and_the_Order_of_the_Phoenix 
------------------------------------------------------

Recommended :  31122.I_Capture_the_Castle
Description :  Through six turbulent months of 1934, 17-year-...
 score :  0.48463490837338163 


Recommended :  18521.A_Room_of_One_s_Own
Description :  A Room of One's Own is an extended essay by Vi...
 score :  0.4233528886559716 


Recommended :  11507.My_Name_Is_Asher_Lev
Description :  Asher Lev is a Ladover Hasid who keeps kosher,...
 score :  0.38974214082538 


Recommended :  119073.The_Name_of_the_Rose
Description :  The year is 1327. Benedictines in a wealthy It...
 score :  0.37289630021529535 


Recommended :  15997.Paradise_Lost
Description :  John Milton's Paradise Lost is one of the grea...
 score :  0.3642583814592007 


Recommended :  3388.Corelli_s_Mandolin
Description :  Captain Corelli’s Mandolin is set in the early...
 score :  0.35543348511330536 


Recommended :  29797.The_Pilgrim_s_Pro


  #  **4.Classification system**

Now , we will attempt to maximize our classification ability bases on descriptions.
Classes will be the genres ( the main one of each book) .


In [ ]:
dataframe13 = dataset1[['bookId', 'description','genreSingle']].copy()


dataframe3=dataframe13.iloc[:books_quantity].copy()

dataframe3.info


<bound method DataFrame.info of                                                  bookId  \
0           2.Harry_Potter_and_the_Order_of_the_Phoenix   
1                            2657.To_Kill_a_Mockingbird   
2                              1885.Pride_and_Prejudice   
3                                        41865.Twilight   
4                                  19063.The_Book_Thief   
5                                    170448.Animal_Farm   
6                        11127.The_Chronicles_of_Narnia   
7                     30.J_R_R_Tolkien_4_Book_Boxed_Set   
8                              18405.Gone_with_the_Wind   
9           386162.The_Hitchhiker_s_Guide_to_the_Galaxy   
10                               370493.The_Giving_Tree   
11                               6185.Wuthering_Heights   
12                                968.The_Da_Vinci_Code   
13                              929.Memoirs_of_a_Geisha   
14                      5297.The_Picture_of_Dorian_Gray   
15    24213.Alice_s_Adve

In [ ]:
####### CONTROL BOX ##################
books_quantity3=1000

######################################



def strip_everything(ww):
  w1 = ww.strip("'-:()[]{}!?$%^&*_~\;/.`+=#@0123456789")
  return w1


stopwords = nltk.corpus.stopwords.words('english')
#stopwords.extend( ['and', 'didnt',])
lemmatizer = WordNetLemmatizer()

# for each row
for jj in  range (len( dataframe3  ) ):


  cour= dataframe3.iloc[jj,1]
  if books_quantity3<51:
    print ( cour,"\n")





  cour = cour.replace("−", " 9 ")
  cour = re.sub(r'-' ,' 9 ',cour);

  ######  Important ** The sequence of our cleaning ,converting actions affects
  #####   the final result.Thus we  have to be carefoul on possible conflicts
  #### i.e. bla bla bla - blu1 blu blu blu2 - ble ble ble
  ### may become bla bla blablu1 ..... blu2ble ble ble
  # instead of bla bla bla blu1 blu blu blu2 ble ble ble


  #####  A regular expression way to delete everything that is not a w ( \w)
  #####  or whitespace (\s).Negation is with  symbol :  ^
  #####  \d for  digiits [0-9]
  #####  | == OR
  #####  \'\w+ for letter after '

  cour = re.sub(r'\'\w+', '', cour)
  cour  =re.sub(r'[^\w\s]|\d', '', cour)

  #a lambda calculus way for the same purpose
  #cour=' '.join(list(map(lambda x: x[:x.find("'")] if "'" in x else x, cour.split(' '))))

  #convert every upper case character to it's lower case
  cour = cour.lower()

  #deleting parenthesis and non alphanumerical characters
  cour = strip_everything(cour)

  #convert text to list/vector
  #cour =word_tokenize(cour)
  if books_quantity3<51:
    print ( cour,"\n")

  """
  filtered_sentence = [w for w in word_tokens if not w.lower() in stopwords]

  for w in word_tokens:
    if w not in stopwords:
        filtered_sentence.append(w)

  """


  lemmatizer= WordNetLemmatizer()
  for kk in range  ( len ( cour)):
   #print ( filtered_sentence[kk])
   #print ( lemmatizer.lemmatize(filtered_sentence[kk]))
   lemmed_result= lemmatizer.lemmatize(cour[kk])

  lemmedresult=textTolemma(cour)


  dataframe3.at[jj,'description']=lemmedresult


dataframe3.head(200)
dataframe3=dataframe3.dropna(subset=['genreSingle'])

In [ ]:
genreList = dataframe3['genreSingle'].tolist()


uniq_genre= []

genre_labels=[]
genrelabel=1

for item in genreList:
    if  (item not in uniq_genre ):
        tem=[item,genrelabel]
        genrelabel+=1
        uniq_genre.append(item)
        genre_labels.append(tem)

print("Genre   |      Label ")
for item in genre_labels:
    print(item)



labelsList=[]

for x in genreList:
  for y in genre_labels:
    if x==y[0]:
      labelsList.append(y[1])


dataframe3['genreLabel']=labelsList

def labeltoclass( labelID,genlabels=genre_labels):
  for i in genlabels:
    if labelID==genlabels[1]:
      return genlabels[0]


#dataframe3.head(20)

Genre   |      Label 
['Fantasy', 1]
['Classics', 2]
['Young Adult', 3]
['Historical Fiction', 4]
['Science Fiction', 5]
['Childrens', 6]
['Fiction', 7]
['Horror', 8]
['Nonfiction', 9]
['Romance', 10]
['Mystery', 11]
['Picture Books', 12]
['Graphic Novels', 13]
['Poetry', 14]
['History', 15]
['Travel', 16]
['Philosophy', 17]
['Science', 18]
['Plays', 19]
['Christian Fiction', 20]
['Short Stories', 21]
['Thriller', 22]
['Mythology', 23]
['Paranormal', 24]
['Christmas', 25]
['Vampires', 26]
['Chick Lit', 27]
['Self Help', 28]
['Urban Fantasy', 29]
['Religion', 30]
['Contemporary', 31]
['Survival', 32]
['Memoir', 33]
['Christian', 34]
['True Crime', 35]
['Comics', 36]
['Biography', 37]


<ipython-input-23-558378b18b15>:30: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe3['genreLabel']=labelsList


In [ ]:
ggg = dataframe3['description'].tolist()

ggg[1]

'the unforgettable novel of a childhood in a sleepy southern town and the crisis of conscience that rocked it to kill a mockingbird became both an instant bestseller and a critical success when it wa first published in it went on to win the pulitzer prize in and wa later made into an academy award winning film also a classiccompassionate dramatic and deeply moving to kill a mockingbird take reader to the root of human behavior to innocence and experience kindness and cruelty love and hatred humor and pathos now with over million copy in print and translated into forty language this regional story by a young alabama woman claim universal appeal harper lee always considered her book to be a simple love story today it is regarded a a masterpiece of american literature'

In [ ]:
descriptions = dataframe3['description'].tolist()




from gensim.models import Word2Vec
import gensim.downloader
from gensim.test.utils import datapath
from gensim.models import KeyedVectors
import scipy
from sklearn.manifold import TSNE

print(list(gensim.downloader.info()['models'].keys()))

['fasttext-wiki-news-subwords-300', 'conceptnet-numberbatch-17-06-300', 'word2vec-ruscorpora-300', 'word2vec-google-news-300', 'glove-wiki-gigaword-50', 'glove-wiki-gigaword-100', 'glove-wiki-gigaword-200', 'glove-wiki-gigaword-300', 'glove-twitter-25', 'glove-twitter-50', 'glove-twitter-100', 'glove-twitter-200', '__testing_word2vec-matrix-synopsis']


Downloading pretrained embeddings.We need a small one .

In [ ]:
##  CONTROL3


## comment1 :: Delete before first execution
#das_vectors=gensim.downloader.load('glove-wiki-gigaword-200')

Loading embedding

[**Otherwise, file will be here for manual download.**](https://drive.google.com/drive/folders/1xSO_oSmIwYewKXcBV1XEUuG7lgzXDu8L?usp=sharing)

In [ ]:
path22='/content/drive/MyDrive/DataMining/subdatamining'




path22= path22+'/fastext.kv'
import gc
import pickle


## comment2 :: Delete before first execution
"""
pick_insert = open(path22,'wb')
pickle.dump(das_vectors, pick_insert)
pick_insert.close()
"""


pick_read = open(path22,'rb')
das_vectors = pickle.load(pick_read,encoding="utf-8")
pick_read.close()





Converting each preproccessed description from string to list of strings .

In [ ]:
coun=0

print ( len ( descriptions))

for indexy in range (len ( descriptions)) :


  splitted=descriptions[indexy].split()
  descriptions[indexy]=splitted

#print( descriptions)



1000


In [ ]:
model131 = KeyedVectors.load(path22)

**Word to vector**

Now we compute a vector of similarity with the 200 -wisely choosed from wiki - features for each word of each description.Right after we find the average value of each description in a same size (200) feature base .
This should suffice for upcoming classification with low demands ( based of course on the tiny runtime and space enviroment we get) .

In [ ]:
first_shape=(  len (descriptions),200 )

class_data1=np.zeros(first_shape)


average_vectors=[]

for i in range (len (descriptions)):
  wordcount=0
  temp_vector=np.zeros(200)
  for word in descriptions[i]:
    wordcount+=1
    if word in model131:
      vector_result=model131[word]

      temp_vector+=vector_result
    else :
      continue


  avgvec=temp_vector/wordcount

  class_data1[i]=avgvec
  average_vectors.append(avgvec)


#print ( class_data1[1])


In [ ]:
print(class_data1.shape)

(1000, 200)


In [ ]:
dataframe3['average200_features']=average_vectors

<ipython-input-32-e4484a8b2b02>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe3['average200_features']=average_vectors


**Naive Bayes**

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

from sklearn.preprocessing import StandardScaler

In [ ]:
X_train, X_test, y_train, y_test =train_test_split(average_vectors,labelsList,test_size= 0.2, random_state=1)

# scalling the input data
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.fit_transform(X_test)

# import Gaussian Naive Bayes classifier
from sklearn.naive_bayes import GaussianNB
# create a Gaussian Classifier
classifer1 = GaussianNB()
# training the model
classifer1.fit(X_train, y_train)
# testing the model
y_pred1 = classifer1.predict(X_test)


from sklearn.metrics import accuracy_score
# printing the accuracy of the model
print(accuracy_score(y_pred1, y_test))

from sklearn.metrics import classification_report
# printing the report
print(classification_report(y_test, y_pred1,zero_division=True))

0.13
              precision    recall  f1-score   support

           1       0.16      0.19      0.17        26
           2       0.30      0.23      0.26        60
           3       0.00      0.00      0.00         8
           4       0.14      0.22      0.17         9
           5       0.00      0.00      0.00         8
           6       0.00      0.00      0.00         1
           7       0.10      0.04      0.06        48
           8       0.00      0.00      0.00         4
           9       0.00      0.00      0.00        10
          10       0.07      0.43      0.12         7
          11       0.00      0.00      0.00         6
          14       1.00      0.00      0.00         3
          15       1.00      0.00      0.00         1
          17       0.00      0.00      0.00         1
          18       1.00      0.00      0.00         2
          24       1.00      0.00      0.00         1
          26       1.00      0.00      0.00         2
          27       1.0

**SVM**

In [ ]:
from sklearn.linear_model import Perceptron, LogisticRegression
from sklearn.svm import SVC
from sklearn import datasets
from sklearn import metrics

In [ ]:
svc = SVC(C=1.0, random_state=54647, kernel='rbf',gamma='auto')

svc.fit(X_train,y_train)

# Make the predictions
y_predict = svc.predict(X_test)

# Measure the performance
print("Kernel == RBF \nAccuracy score %.3f" %metrics.accuracy_score(y_test, y_predict))


# printing the accuracy of the model
print(accuracy_score(y_predict, y_test))


# printing the report
print(classification_report(y_test, y_predict,zero_division=True))

Kernel == RBF 
Accuracy score 0.250
0.25
              precision    recall  f1-score   support

           1       0.50      0.04      0.07        26
           2       0.30      0.23      0.26        60
           3       1.00      0.00      0.00         8
           4       1.00      0.00      0.00         9
           5       1.00      0.00      0.00         8
           6       1.00      0.00      0.00         1
           7       0.23      0.73      0.35        48
           8       1.00      0.00      0.00         4
           9       1.00      0.00      0.00        10
          10       1.00      0.00      0.00         7
          11       1.00      0.00      0.00         6
          14       1.00      0.00      0.00         3
          15       1.00      0.00      0.00         1
          17       1.00      0.00      0.00         1
          18       1.00      0.00      0.00         2
          24       1.00      0.00      0.00         1
          26       1.00      0.00      0

In [ ]:
svc = SVC(C=1.0, random_state=54647, kernel='linear',gamma='auto')

svc.fit(X_train,y_train)

# Make the predictions
y_predict = svc.predict(X_test)

# Measure the performance
print("Kernel == LINEAR \nAccuracy score %.3f" %metrics.accuracy_score(y_test, y_predict))


# printing the accuracy of the model
print(accuracy_score(y_predict, y_test))


# printing the report
print(classification_report(y_test, y_predict,zero_division=True))

Kernel == LINEAR 
Accuracy score 0.140
0.14
              precision    recall  f1-score   support

           1       0.17      0.27      0.21        26
           2       0.25      0.18      0.21        60
           3       0.00      0.00      0.00         8
           4       0.00      0.00      0.00         9
           5       0.00      0.00      0.00         8
           6       0.00      0.00      0.00         1
           7       0.22      0.21      0.22        48
           8       0.00      0.00      0.00         4
           9       0.00      0.00      0.00        10
          10       0.00      0.00      0.00         7
          11       0.00      0.00      0.00         6
          14       1.00      0.00      0.00         3
          15       0.00      0.00      0.00         1
          17       1.00      0.00      0.00         1
          18       1.00      0.00      0.00         2
          19       0.00      1.00      0.00         0
          24       1.00      0.00    

**Random Forests**  


In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV, train_test_split
from scipy.stats import randint

# Tree Visualisation
from sklearn.tree import export_graphviz
from IPython.display import Image
import graphviz



rf = RandomForestClassifier()
rf.fit(X_train, y_train)

rf_y_pred = rf.predict(X_test)

# printing the accuracy of the model
print(accuracy_score(rf_y_pred, y_test))


# printing the report
print(classification_report(y_test,rf_y_pred,zero_division=True))

0.22
              precision    recall  f1-score   support

           1       0.25      0.08      0.12        26
           2       0.26      0.25      0.26        60
           3       0.00      0.00      0.00         8
           4       1.00      0.00      0.00         9
           5       1.00      0.00      0.00         8
           6       1.00      0.00      0.00         1
           7       0.20      0.56      0.30        48
           8       1.00      0.00      0.00         4
           9       1.00      0.00      0.00        10
          10       1.00      0.00      0.00         7
          11       1.00      0.00      0.00         6
          14       1.00      0.00      0.00         3
          15       1.00      0.00      0.00         1
          17       1.00      0.00      0.00         1
          18       1.00      0.00      0.00         2
          24       1.00      0.00      0.00         1
          26       1.00      0.00      0.00         2
          27       1.0